runtime/src/chpl-gpu.c

```diff
@@ -253,6 +286,12 @@ void* chpl_gpu_mem_realloc(void* memAlloc, size_t size,
   assert(chpl_gpu_is_device_ptr(memAlloc));
 
   size_t cur_size = chpl_gpu_get_alloc_size(memAlloc);
+
+  // this is not really cool. Should we have a realloc in the impl layer?
+  if (cur_size == -1) {
+    return chpl_mem_realloc(memAlloc, size, description, lineno, filename);
```

Comment:
Why did you add this?

Reply:
So, this is one of the earliest changes, and we probably have a better way of doing this right now.

The story goes: there is no reallocation in the CUDA API (probably neither in HIP). Here, this function does it in a home-grown way. It involves creating a new allocation, copying the old allocation into it and freeing the old one. But copying the old allocation requires a priori knowledge of the size of the old allocation. What CUDA API has instead is the ability to query that for a device pointer.

However, we don't have that ability available to us in this mode, so I made `chpl_gpu_get_alloc_size` to return -1 as a sentinel instead. Upon receiving that sentinel, we fall back to the normal memory layer. But maybe, we can check `GPU_RUNTIME_CPU` instead, now that I am thinking about this.